### PR TITLE
Add methods to get next alarm datetime

### PR DESCRIPTION
**Pull request by DaAwesomeP**
I have added methods to more easily determine:
- The next `datetime` for an `Alarm`
- The next `datetime` for an `Alarms` (i.e. a whole zone/room)

These methods are particularly useful for home automation scenarios.

~~This is still a draft because I have not tested it yet.~~

Maybe there is a better name for the `bypass_disabled` argument?

### CODE REVIEW

**Comment by pwt** [comment]
Thanks for the contribution.

> Maybe there is a better name for the bypass_disabled argument?

I'd be minded to invert this to be `include_disabled`, and to make the default `False`.

File: soco/alarms.py [hunk 1]
```diff
@@ -179,6 +185,35 @@ def update(self, zone=None):
...
+                to `datetime().now()`.
+            bypass_disabled (bool, optional): If `True` then disabled alarms
+                will be included in searching for the next alarm. Defaults to
+                `False`.
+
+        Returns:
+            datetime: The next alarm trigger datetime or None if disabled
+        """
+        if from_datetime is None:
+            from_datetime = datetime().now()
```
Comment (pwt) [inline soco/alarms.py:205]: This should be `datetime.now()`

File: soco/alarms.py [hunk 2]
```diff
@@ -372,6 +407,71 @@ def alarm_id(self):
...
+                `False`.
+
+        Returns:
+            datetime: The next alarm trigger datetime or None if disabled
+        """
+        if not self.enabled and not bypass_disabled:
+            return None
+
+        if from_datetime is None:
+            from_datetime = datetime().now()
```
Comment (pwt) [inline soco/alarms.py:430]: Replace with `datetime.now()`

**Comment by pwt** [comment]
I've fixed the check jobs so that all checks on `master` are now passing. If you rebase, you'll pick up the fix.

You'll also need to address the following -- see code review comments.
```
************* Module soco.alarms
soco/alarms.py:205:28: E1120: No value for argument 'year' in constructor call (no-value-for-parameter)
soco/alarms.py:430:28: E1120: No value for argument 'year' in constructor call (no-value-for-parameter)
soco/alarms.py:470:35: E1101: Class 'datetime' has no 'timedelta' member (no-member)
```

File: soco/alarms.py (hunk 1 above)
Comment (DaAwesomeP) [reply soco/alarms.py]: Fixed!

File: soco/alarms.py (hunk 2 above)
Comment (DaAwesomeP) [reply soco/alarms.py]: Fixed!

File: soco/alarms.py [hunk 3]
```diff
@@ -372,6 +407,71 @@ def alarm_id(self):
...
+
+        # Find first day
+        from_datetime_day = from_datetime.weekday()
+        offset_weekday = (from_datetime_day + offset) % 7
+        while offset_weekday not in recurrence_set:
+            offset += 1
+            offset_weekday = (from_datetime_day + offset) % 7
+
+        return datetime.combine(
+            from_datetime.date() + datetime.timedelta(days=offset),
```
[the remark this answers is not in the document]
Comment (DaAwesomeP) [reply soco/alarms.py]: Fixed!

**Comment by DaAwesomeP** [comment]
@pwt Thanks for the review! OK, I have fixed your notes (I did not see those errors for Python 3.12 before). I have also now actually tried this out and fixed some other bugs that I did not find before. I added a `zone_uid` filter to make it easier to for the next alarm by device zone.

**Review by pwt — CHANGES_REQUESTED** [review]
This all looks great to me -- thanks.

I'd remove the `assert` statements (lines 453 and 461/462), unless there's a specific reason you think they should remain for the production release -- then I think we're all set to merge.

**Comment by DaAwesomeP** [comment]
> I'd remove the `assert` statements (lines 453 and 461/462), unless there's a specific reason you think they should remain for the production release -- then I think we're all set to merge.

Ah, yeah sorry, all fixed! Thank you!

**Review by pwt — APPROVED** [review]

**Pull request merged**